photon-lib/src/main/scala/com/linkedin/photon/ml/util/Implicits.scala

```diff
@@ -0,0 +1,106 @@
+   *         val outputDir = summarizationOutputDir + "/" + featureShardId
+   *         val indexMap = indexMapLoaders(featureShardId).indexMapForDriver()
+   *         IOUtils.writeBasicStatistics(sc, shardSummary, outputDir, indexMap)
+   *       }
+   *
+   * Without having to create a val, then calling map on the val, then repeating the val name at the end
+   * to do something with it. In essence, we append a block of side-effect'ing code to a chain of operations
+   * on a Map.
+   *
+   * @param m A Map to tap
+   * @tparam K The type of the keys in Map m
+   * @tparam V The type of the values in Map m
+   */
+  implicit class TapMap[K, V](m: Map[K, V]) {
+
+    def tap(f: ((K, V)) => Unit): Map[K, V] = { m.foreach(f); m }
+    def tapKeys(f: K => Unit): Map[K, V] = { m.keysIterator.foreach(f); m }
+    def tapValues(f: V => Unit): Map[K, V] = { m.valuesIterator.foreach(f); m }
+  }
+
+  /**
+   * Tap for Set, List, Iterable ... Same idea as tap for Map, same usage pattern: append some side effect in
+   * a chain of operations on a Set, List ...
+   *
+   * @param c A Collection to iterate on
+   * @param ev An implicit to make sure the collection is TraversableOnce
+   * @tparam Coll The type of the Collection
+   * @tparam T The type of the elements in the Collection
+   */
+  implicit class TapTraversable[Coll[_], T](c: Coll[T])(implicit ev: Coll[T] <:< TraversableOnce[T]) {
+    def tap(f: T => Unit): Coll[T] = { ev(c).foreach(f); c }
+  }
+
+  /**
+   * Tap for Array, which is not quite Iterable. See documentation for TapMap.
+   *
+   * @param a The Array to tap
+   * @tparam T The type of whatever is in the Array
+   */
+  implicit class TapArray[T](a: Array[T]) {
+    def tap(f: T => Unit): Array[T] = { a.foreach(f); a }
+  }
+
+  /**
+   * Tap for Options. See documentation for TapMap.
+   *
+   * @param o An Option to tap
+   * @tparam T The type of the thing contained in the Option
+   */
+  implicit class TapOption[T](o: Option[T]) {
+    def tap(f: T => Unit): Option[T] = { o.foreach(f); o }
+  }
+
+  /**
+   * Extract a value from a Map 'through' an Option. This allows to write simpler code, e.g.:
+   *
+   *   normalizationContexts.extractOrElse(shardId)(defaultNormalizationContext)
+   *
+   * given:
+   *
+   *   normalizationContexts: Option[Map[String, NormalizationContext]]
+   *
+   * When an Option contains a Map[K, V], extract V at K key from Map if Option isDefined, else apply f that
+   * returns a default V.
+   *
+   * NOTE: throws an exception if the Option isDefined but key is not found. The behavior of the Map is not changed,
+   * the default provided is used only in case the Option itself isEmpty.
+   *
+   * @param o An Option that contains a Map[K, V]
+   * @tparam K The type of the keys in the Map
+   * @tparam V The type of the values in the Map
+   */
+  implicit class ExtractOrElse[K, V](o: Option[Map[K, V]]) {
+    def extractOrElse(key: K)(f: => V): V = { if (o.isDefined) o.get(key) else f }
```

Comment:
More idiomatic:

```scala
o.flatMap(_.get(key)).getOrElse(f)
```

Also, why not just use this form at the call site instead of the wrapper function?